execution/src/vm.rs

```diff
@@ -98,48 +38,141 @@ pub struct ExecutionContext {
     pub call_stack: VecDeque<Address>,
 }
 
-#[derive(WasmerEnv, Clone)]
-pub struct SharedExecutionContext(pub Arc<Mutex<ExecutionContext>>);
+#[derive(Clone)]
+/// A map of modules, a shared u32, and a list of imports.
+pub struct SharedExecutionContext(pub Arc<Mutex<(AddressHashMap<Module>, u32, Vec<Extern>)>>);
 
 pub struct VM {
     cfg: ExecutionConfig,
-    imports: ImportObject,
-    store: Store,
+    imports: Instance,
     final_ledger: Arc<Mutex<SCELedger>>,
     step_history: VecDeque<(Slot, Option<BlockId>, SCELedgerChanges)>,
     current_execution_context: SharedExecutionContext,
 }
 
 impl VM {
     pub fn new(cfg: ExecutionConfig) -> VM {
-        let store = Store::default();
+        // Execute a contract that will:
+        // 1. Call into another contract, and
+        // 2. Mutate some shared state.
+        let engine = Engine::default();
         let final_ledger = Arc::new(Mutex::new(
             SCELedger::default(), // TODO bootstrap
         ));
-        let current_execution_context =
-            SharedExecutionContext(Arc::new(Mutex::new(ExecutionContext {
-                ledger_step: SCELedgerStep {
-                    final_ledger: final_ledger.clone(),
-                    cumulative_history_changes: Default::default(),
-                    caused_changes: Default::default(),
-                },
-                max_gas: Default::default(),
-                coins: Default::default(),
-                gas_price: Default::default(),
-                slot: Slot::new(0, 0),
-                opt_block_id: Default::default(),
-                opt_block_creator_addr: Default::default(),
-                call_stack: Default::default(),
-            })));
-        let imports = imports! {
-            "env" => {
-                "foo" => Function::new_native_with_env(&store, current_execution_context.clone(), foo),
-            },
+
+        let current_execution_context = {
+            // Add a module for the address "hello world".
+            // The module will call `host_set` to mutate the shared u32.
+            let mut modules: AddressHashMap<Module> = Default::default();
+            let wat = r#"
+            (module
+                (import "host" "set" (func $host_set (param i32)))
+                (import "host" "call" (func $host_call (param i32 i32)))
+                (func (export "main")
+                    i32.const 3
+                    call $host_set)
+            )
+        "#;
```

Comment:
where this code come from? the style of arguments pointer+size isn't standardized if you want to develop in c++, grain etc.

Reply:
I copied it from various examples found in the docs: https://docs.rs/wasmtime/latest/wasmtime/